sklearn/metrics/_dist_metrics.pyx.tp

```diff
@@ -445,24 +700,20 @@ cdef class DistanceMetric{{name_suffix}}:
             The shape (Nx, Ny) array of pairwise distances between points in
             X and Y.
         """
-        cdef cnp.ndarray[{{DTYPE_t}}, ndim=2, mode='c'] Xarr
-        cdef cnp.ndarray[{{DTYPE_t}}, ndim=2, mode='c'] Yarr
-        cdef cnp.ndarray[{{DTYPE_t}}, ndim=2, mode='c'] Darr
+        Y = X if Y is None else Y
 
-        Xarr = np.asarray(X, dtype={{DTYPE}}, order='C')
-        self._validate_data(Xarr)
-        if Y is None:
-            Darr = np.zeros((Xarr.shape[0], Xarr.shape[0]),
-                         dtype={{DTYPE}}, order='C')
-            self.pdist(Xarr, Darr)
-        else:
-            Yarr = np.asarray(Y, dtype={{DTYPE}}, order='C')
-            self._validate_data(Yarr)
-            Darr = np.zeros((Xarr.shape[0], Yarr.shape[0]),
-                         dtype={{DTYPE}}, order='C')
-            self.cdist(Xarr, Yarr, Darr)
-        return Darr
+        X_is_sparse = issparse(X)
+        Y_is_sparse = issparse(Y)
 
+        if not X_is_sparse and not Y_is_sparse:
+            return self._pairwise_dense_dense(X, Y)
+        if X_is_sparse and Y_is_sparse:
+            return self._pairwise_sparse_sparse(X, Y)
+        if X_is_sparse and not Y_is_sparse:
+            return self._pairwise_sparse_dense(X, Y)
+        if not X_is_sparse and Y_is_sparse:
+            # Swapping argument and transposing the result
+            return self._pairwise_sparse_dense(Y, X).T
```

Comment:
An alternative that I've tested is to provide the order to use for the distance matrix (which would have been "F" in this case) so that the result is always "C"-ordered. Yet, the Cython methods expect the distance matrix to be "C"-ordered.

I think the best is simply to implement the last case, as done in 4f4583934c151051510ba7cead1256bb6a7dabe6.